File: rubygems/pray-cli/lib/pray/manifest_constraint.rb
# frozen_string_literal: true

module Pray
  module ManifestMethods
    module_function

    PACKAGE_KEYWORDS = %w[pray use include agent package].freeze

    def rewrite_constraint_on_line(line, constraint)
      indent = line[/\A\s*/]
      trimmed = line.lstrip
      after_keyword = skip_package_keyword(trimmed)
      raise Error.manifest("package declaration is missing a keyword") unless after_keyword

      after_keyword = after_keyword.lstrip
      parsed_name = parse_quoted(after_keyword)
      raise Error.manifest("package declaration is missing a quoted name") unless parsed_name

      name, after_name = parsed_name
      quoted_constraint = "\"#{constraint}\""
      keyword_and_name = trimmed[0, trimmed.length - after_name.length]
      remainder = after_name.lstrip
      return "#{indent}#{keyword_and_name}, #{quoted_constraint}" if remainder.empty?
      unless remainder.start_with?(",")
        raise Error.manifest("package #{name} declaration is missing a comma after the name")
      end

      after_comma = remainder[1..].lstrip
      if after_comma.start_with?("\"", "'")
        parsed_constraint = parse_quoted(after_comma)
        unless parsed_constraint
          raise Error.manifest("package #{name} declaration has an unclosed constraint")
        end

        return "#{indent}#{keyword_and_name}, #{quoted_constraint}#{parsed_constraint[1]}"
      end

      "#{indent}#{keyword_and_name}, #{quoted_constraint}, #{after_comma}"
    end

    def skip_package_keyword(input)
      PACKAGE_KEYWORDS.each do |keyword|
        next unless input.start_with?(keyword)

        rest = input[keyword.length..]
        next_character = rest[0]
        whitespace_or_quote = next_character.nil? ||
          next_character.match?(/\s/) ||
          next_character == "\"" ||
          next_character == "'"
        return rest if whitespace_or_quote
      end
      nil
    end

    def parse_quoted(input)
      quote = input[0]
      return unless quote == "\"" || quote == "'"

      rest = input[1..]
      ending = rest.index(quote)
      return unless ending

      [rest[0, ending], rest[(ending + 1)..]]
    end
  end
end

File: crates/pray-core/src/manifest_constraint.rs
use crate::{PrayError, PrayResult};

const PACKAGE_KEYWORDS: [&str; 5] = ["pray", "use", "include", "agent", "package"];

pub fn rewrite_constraint_on_line(line: &str, constraint: &str) -> PrayResult<String> {
    let indent_len = line.len() - line.trim_start().len();
    let indent = &line[..indent_len];
    let trimmed = line.trim_start();
    let after_keyword = skip_package_keyword(trimmed).ok_or_else(|| {
        PrayError::Manifest("package declaration is missing a keyword".to_string())
    })?;
    let after_keyword = after_keyword.trim_start();
    let (name, after_name) = parse_quoted(after_keyword).ok_or_else(|| {
        PrayError::Manifest("package declaration is missing a quoted name".to_string())
    })?;
    let quoted_constraint = format!("\"{constraint}\"");
    let keyword_and_name = &trimmed[..trimmed.len() - after_name.len()];
    let remainder = after_name.trim_start();
    if remainder.is_empty() {
        return Ok(format!("{indent}{keyword_and_name}, {quoted_constraint}"));
    }
    if !remainder.starts_with(',') {
        return Err(PrayError::Manifest(format!(
            "package {name} declaration is missing a comma after the name"
        )));
    }
    let after_comma = remainder[1..].trim_start();
    if after_comma.starts_with('"') || after_comma.starts_with('\'') {
        let (_, after_constraint) = parse_quoted(after_comma).ok_or_else(|| {
            PrayError::Manifest(format!(
                "package {name} declaration has an unclosed constraint"
            ))
        })?;
        return Ok(format!(
            "{indent}{keyword_and_name}, {quoted_constraint}{after_constraint}"
        ));
    }
    Ok(format!(
        "{indent}{keyword_and_name}, {quoted_constraint}, {after_comma}"
    ))
}

fn skip_package_keyword(input: &str) -> Option<&str> {
    for keyword in PACKAGE_KEYWORDS {
        if let Some(rest) = input.strip_prefix(keyword) {
            if rest.starts_with(char::is_whitespace)
                || rest.starts_with('"')
                || rest.starts_with('\'')
            {
                return Some(rest);
            }
        }
    }
    None
}

fn parse_quoted(input: &str) -> Option<(&str, &str)> {
    let quote = input.chars().next()?;
    if quote != '"' && quote != '\'' {
        return None;
    }
    let rest = input.get(quote.len_utf8()..)?;
    let end = rest.find(quote)?;
    Some((&rest[..end], &rest[end + quote.len_utf8()..]))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn replaces_existing_constraint_and_keeps_keywords() {
        let line = r#"  pray "sample/base", "~> 1.0", export: "testing-basics""#;
        let rewritten = rewrite_constraint_on_line(line, "~> 1.1").expect("rewrite");
        assert_eq!(
            rewritten,
            r#"  pray "sample/base", "~> 1.1", export: "testing-basics""#
        );
    }

    #[test]
    fn inserts_constraint_before_keyword_arguments() {
        let line = r#"    pray "sample/base", path: "packages/base""#;
        let rewritten = rewrite_constraint_on_line(line, "~> 1.1").expect("rewrite");
        assert_eq!(
            rewritten,
            r#"    pray "sample/base", "~> 1.1", path: "packages/base""#
        );
    }

    #[test]
    fn appends_constraint_when_the_line_has_only_a_name() {
        let line = r#"pray "sample/base""#;
        let rewritten = rewrite_constraint_on_line(line, "~> 1.1").expect("rewrite");
        assert_eq!(rewritten, r#"pray "sample/base", "~> 1.1""#);
    }
}
